### ISSUE TITLE
Test Suite Integration

### ISSUE THREAD

**Issue by aendeavor** [test-org-dm/docker-mailserver#1]
## Is your feature request related to a problem?

Currently, no tests are run.

## Describe the solution you'd like

Integrate the Bats test suite and all linting targets into GH actions.

## Alternatives you've considered

Let TravisCI run them. Although I'm not a big fan of this.

## What are you going to contribute?

Help with understanding how the test suite should work.

**Comment by radicand** [comment]
I will take a look into this over the weekend

**Comment by radicand** [comment]
I've started work, can be tracked in #2

**Comment by wernerfred** [comment]
I have a few additional ideas as well as i would like to propose an overview diagram as @aendeavor mentioned (or at least show what's possible). Hopefully I will find time tomorrow. Keep you updated here.

**Comment by aendeavor** [comment]
That‘s absolutely great news @wernerfred! Thank you very much:) I had hoped for this, as this could bring in structure which is very much needed, and I‘m glad you‘ll do it.

Looking forward to hearing from you.

**Comment by radicand** [comment]
Sounds great - feel free to jump on the gha/test-suite branch; I've separated out build and test - primarily working on getting the test suite to fully pass. There are some glitches on image/container names in some of the tests currently I'm attempting to debug. If the existing tests should be broken out, that's an opportunity, and I also do not have a step to push the passed build as the "latest" tag, but that needs to be done as well.

**Comment by wernerfred** [comment]
So here I am with a first update. I was thinking the whole time how to build the setup and which event should trigger which workflow. See this as a fist draft and valuable for further discussions.

The first problem i encountered with gh actions (in comparison to other ci/cd tools) is that you cannot trigger workflows dependant on the status of combination of different other workflows. Although it might exist some workarounds with the [workflow_run hook](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#workflow_run)

So i tried to design a system which runs only the minimum of actions necessary on a given event. I came up with a few assumptions (which might be wrong - i'm not that deep into how releasing new versions worked till now on docker-mailserver - then correct me pls). I identified 3 main events that can trigger a set of actions. 

1. Pull Request (created or updated)
2. Push (on master branch)
3. Release (published)

The 1st event (PR) will trigger 4 different actions:

1. Hadolint (only if Dockerfile was changed)
2. shellcheck (only if .sh files were changed)
3. eclint
4. build + generate-accounts & tests (only test that needs a builded image)

The 2nd event (Push on master) will trigger the same 4 actions and additionally if a tag of the form ```v*.*.*``` is pushed a new release is created. We should think about further jobs that will build the changelog from commit messages (or edit release notes afterwards)

The 3rd event (Release published) will trigger a action that builds the image, tags it accordingly and pushes it to dockerhub. Note: no repetitive testing is needed here as all previous commits would be already validated through one of the other actions. It is not possible to commit sth that is not tested. Now as I'm writing this i came up with the idea that we should exclude changes to '**.md' and other files of this type from triggering actions as they have no impact on the build.

![ghactions](https://user-images.githubusercontent.com/20406381/101396243-532e6c80-38cb-11eb-88b2-614e8537867e.jpg)

Further ideas: use docker containers for actions whereever possible. For example use the hadolint container (```docker run --rm -i hadolint/hadolint < Dockerfile```). There might be separate actions for such tasks, too like: [hadolint-action](https://github.com/burdzwastaken/hadolint-action). This might be easier for contributors to test at their own setup without the need of installing and configuring tools except of docker (which they obviously have if they use docker-mailserver).

So this was my short brainstorming. Did I miss something obvious? Feel free to add your opinion on this. Together we will get further!

**Comment by aendeavor** [comment]
I've got only one thing to say: brilliant. That is **_exactly_** what I wanted to see :D

Now everyone can adhere to these guidelines. I would like to see this in a file called `TESTING.md` in the root of this repo.

---

> Now as I'm writing this i came up with the idea that we should exclude changes to '**.md' and other files of this type from triggering actions as they have no impact on the build.

I agree, this is a good idea.

--- 

> Further ideas: use docker containers for actions whereever possible. For example use the hadolint container (docker run --rm -i hadolint/hadolint < Dockerfile). There might be separate actions for such tasks, too like: hadolint-action. 

I agree again.

---

What's now to do: Because this is A LOT of work, what would now be necessary is having a bare-minimum implementation of you actions-plan @wernerfred. You would not need to implement every detail, but since you came of with this (absolutely brilliant) piece (of art), you are perfect for giving everyone a template to work with, so that we are all on the same page.

What do you think?

**Comment by wernerfred** [comment]
Thanks a lot 👍🏻 
I will definitely provide all action base files but as you mentioned they need to be fine-tuned then. But first off all i would like to wait for @radicand to reply/react to my proposal as he did already some work (which i can reuse for sure). We need to clarify where i should start working (new branch or @radicands branch?) and i will need some additional support from time to time to understand the logic of certain parts in the code i think.

How do you all think of the proposal that pushing a tag to the master will trigger a release to be created (we could think about using squashing merges at PRs and the squash commit message follows the [conventional commit guidelines](https://www.conventionalcommits.org/en/v1.0.0/). There are actions that can 
1. Check for compliance with the specs ([action](https://github.com/agenthunt/conventional-commit-checker-action))
2. Build a release description changelog based on the commits ([action](https://github.com/TriPSs/conventional-changelog-action))

Sadly I discovered a disadvantage of my plan. Currently all checks that are triggered on a PR would be triggered again if the PR is merged into master. But I might have already a solution (```github.event.pull_request.merged == false``` on master push triggers). I will need to take a closer look and update the chart.

**Comment by aendeavor** [comment]
That's a good idea. As I will be a bit shorter on time for the rest of this week, it's probably best to coordinate with @radicand.

The proposal for pushing with a tag is a good idea, but I don't know whether this is realized in practice. I would do this in the end rather than in the beginning.

**Comment by radicand** [comment]
Thanks all - happy to have you fork the branch off of mine and do your hacking.

Your diagram seems like a good start to me - I'd make Test dependent on the previous 3 linter steps though; maybe that's the intention, but the diagram seems to indicate they are parallel activities.

Additionally, if the test/lint has to run again on merges, I don't think it's the end of the world (though it sounds like we have a solution here).

Lastly on tagging, I'd love to see if there's some action/automation for managing the tags. I know there's stuff we can do for review/approve to automate workflows and merging, I've got to imagine there's a way to automate the tagging and release with tags as well.

It's also worth noting that on the test suite step, I'm running into some very weird failures I can't replicate outside of the GHA environment (e.g., works for me locally). I'm wondering if we're hitting some undocumented limit in the amount of DIND containers we can create in one VM, as all the tests after a certain point fail with docker errors. If there's a way to parallelize the tests to run a smaller set in parallel, this might work around it (if it is indeed the issue), alternatively, having some cleanup step (again, guessing) that removes unused containers might help.

**Comment by aendeavor** [comment]
> Your diagram seems like a good start to me - I'd make Test dependent on the previous 3 linter steps though; maybe that's the intention, but the diagram seems to indicate they are parallel activities.

I think so too.

---

> It's also worth noting that on the test suite step, I'm running into some very weird failures I can't replicate outside of the GHA environment (e.g., works for me locally). I'm wondering if we're hitting some undocumented limit in the amount of DIND containers we can create in one VM, as all the tests after a certain point fail with docker errors. If there's a way to parallelize the tests to run a smaller set in parallel, this might work around it (if it is indeed the issue), alternatively, having some cleanup step (again, guessing) that removes unused containers might help.

I will mention @martin-schulze-vireso. Maybe he has an idea what's up with that:)

**Comment by martin-schulze-vireso** [comment]
> I'm running into some very weird failures I can't replicate outside of the GHA environment (e.g., works for me locally).

Could you show instances of these failures? I looked at some of the failures on #2 but there was no "all tests fail from a certain point" problem. Rather, all tests of the setup.sh failed. I'd say this is a problem with the PR, so I commented over there.

> If there's a way to parallelize the tests to run a smaller set in parallel, this might work around it (if it is indeed the issue)

I am not sure I understand what you want to achieve. You can filter down bats test suites e.g. by calling `bats` on fewer files at a time. With newer versions you can also use `-f` to filter on test names and add `-j` to run tests in parallel. However, I don't see how any of this would help here.

> alternatively, having some cleanup step (again, guessing) that removes unused containers might help

The `teardown_file` function should actually take care of the created containers. I am not sure if there are leftover containers that were created within tests. If you feel there are leftover containers, you should look at `docker ps` after the test run.

**Comment by radicand** [comment]
I believe I've narrowed down the issue to tests that invoke `setup.sh`, it appears the container name is not getting passed or set. Will be working to get to the bottom of it.

**Comment by wernerfred** [comment]
Hm i though i posted a reply but can't find it. Anyway here is it again:

> Your diagram seems like a good start to me - I'd make Test dependent on the previous 3 linter steps though; maybe that's the intention, but the diagram seems to indicate they are parallel activities.

Hey @radicand  I would like to make them dependent as well. But as far as I know it is not possible to make workflows dependent on other workflow runs, isn't it? You can make dependencies (or lets call it requirements) between different jobs (```needs:```). I am not even sure if this works between stages.  So if you know sth. regarding to that let me know. The only option then would be to throw them all into one workflow file and make them dependant (only if needs: works for stages as well). So technically the diagram shows the correct plan (parallel) although it would be great if we could connect them

Basically I would like to split the workflow files in as many separate ones as possible to provide a better control (triggers), enhanced readability and easier maintainability even for "new" contributors. 

What do you think

**Comment by radicand** [comment]
> > 
> 
> Hey @radicand I would like to make them dependent as well. But as far as I know it is not possible to make workflows dependent on other workflow runs, isn't it? You can make dependencies (or lets call it requirements) between different jobs (`needs:`). I am not even sure if this works between stages. So if you know sth. regarding to that let me know. The only option then would be to throw them all into one workflow file and make them dependant (only if needs: works for stages as well). So technically the diagram shows the correct plan (parallel) although it would be great if we could connect them

I did some testing using dependant jobs within the workflow (I don't believe we need different workflows here), and it works well. You can see it in action here: https://github.com/test-org-dm/docker-mailserver/actions/runs/417855095 (notice the linter job) -- it failed in this run https://github.com/test-org-dm/docker-mailserver/actions/runs/417828120 and aborted the rest as intended.

**Comment by wernerfred** [comment]
Yeah i know how it works on job level. But what i want to achieve is that runs are only triggered if they are useful. E.g. running hadolint only if the Dockerfile changes etc. This can only be done by different triggers and therefore we need different workflow files.

There is no need to run specific actions when it is obvious from the beginning that nothing changed in the related files. Therefore the sentence "but the diagram seems to indicate they are parallel activities." is only partially correct because in fact they are parallel but only if you change as many files that require **all** linting tests to run (Changes to Dockerfile, at least 1 sh and at least one other file).

Furthermore the build can succeed even if linting tests fail (then the code might not be formatted as we want it but it would be functional). Here it is important to state that the action does only a build. No image is pushed (which might contain not well formatted code). Pushing the image is done in a different action. I'm still looking into possibilities to trigger workflows from other workflows then we could for example say before doing a push everything needs to be run again without errors. This  behavior would be a huge benefit regarding pipeline run times etc.  

This should be a general discussion as it decides mainly on how to structure the actions. @aendeavor what is your opinion on that?

**Comment by martin-schulze-vireso** [comment]
> Can someone familiar with the test suite review https://github.com/test-org-dm/docker-mailserver/runs/1543452144?check_suite_focus=true and review the test run log to propose ideas?

Looking over that I am puzzled why the container name is empty. The normal control flow looks about right. For debugging it would be helpful to see the exact command being try to execute on setup.sh:186. Could you just echo it before execution? From there I'd work backwards when the container name became empty.

**Comment by martin-schulze-vireso** [comment]
I tried to look into this a bit more and created a PR with these changes on my fork. Even after setting up the dockerhub credentials as secrets, I still don't get to the test stage because the [build fails](https://github.com/martin-schulze-vireso/docker-mailserver/pull/2/checks): 

```
#68 [linux/arm64 22/42] RUN su - amavis -c "razor-admin -create && razor-adm...
#68 23.45 register: Error reading socket
#68 ERROR: executor failed running [/bin/bash -o pipefail -c su - amavis -c "razor-admin -create && razor-admin -register"]: exit code: 2
------
 > [linux/arm64 22/42] RUN su - amavis -c "razor-admin -create && razor-admin -register":
------
failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/bash -o pipefail -c su - amavis -c "razor-admin -create && razor-admin -register"]: exit code: 2
Error: buildx call failed with: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/bash -o pipefail -c su - amavis -c "razor-admin -create && razor-admin -register"]: exit code: 2
```

**Comment by radicand** [comment]
@martin-schulze-vireso thanks for the link to the act repo - I'll check it out. Regarding your build failure, I'd suggest attempting it again -- it hasn't happened recently to me, but sometimes I've had the build randomly error out but be find on another execution.

**Comment by radicand** [comment]
I finally got to the bottom of it I believe, thanks @martin-schulze-vireso for the prompt to echo out the attempted command. In GHA, `tty -s` returns a failure, and so USE_TTY ends up being empty. As a result, the docker exec command generated has an extra quoted empty place in it, which the docker executable is treating as the container name. Quick local reproduction: 

```sh
❯ docker exec -ti "" 477 sh
Error: No such container: 
```

Which is the same error we are seeing on GHA. https://github.com/test-org-dm/docker-mailserver/runs/1601855166?check_suite_focus=true is running currently, but already passing previous test failure points. I'll update the thread when it completes.

**Comment by martin-schulze-vireso** [comment]
Nice to see it progress again. I'll unsubscribe now, ping me back in, if there is another problem.

**Comment by radicand** [comment]
Tests pass now 🎉, will merge this into the main branch I'm working on.

**Comment by radicand** [comment]
The E2E lint/build/test pipeline now passes finally. Last thing I'll attempt is to get the image tagging to happen after tests pass only.

**Comment by aendeavor** [comment]
Greak work! I was not reading all since I had some other work and some issues in the main repository, but this is great work!

**Comment by aendeavor** [comment]
@radicand reach back to me when you're done. It seems as if Travis is slowly dying in the main repo so cannot merge PRs any longer. Therefore we will need Gh Actions in the new repo as fast as possible.

**Comment by radicand** [comment]
@aendeavor it's all set, I forgot to update this issue when i finished it. What will happen is each commit will result in a Docker build that gets pushed with the SHA as its tag, and the test suite runs against that. If the test suite passes, the image is re-tagged with the appropriate human-readable tag - e.g., `edge`, `nightly`, `latest`, or a `v#.#.#` tag if the commit is tagged that way. If that sounds good, it's good to use as-is.

**Comment by aendeavor** [comment]
> @aendeavor it's all set, I forgot to update this issue when i finished it. What will happen is each commit will result in a Docker build that gets pushed with the SHA as its tag, and the test suite runs against that. If the test suite passes, the image is re-tagged with the appropriate human-readable tag - e.g., `edge`, `nightly`, `latest`, or a `v#.#.#` tag if the commit is tagged that way. If that sounds good, it's good to use as-is.

That's perfect.

**Comment by aendeavor** [comment]
@radicand could you provide a PR with these changes for the new repository? I've mentioned this repo in the main tracking issue in tomav/docker-mailserver.

I will post a link as soon as I'm not on mobile anymore.

**Comment by radicand** [comment]
Sounds good - will do.
